clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2797,19 +2764,62 @@ void handleResponse(AbstractResponse abstractResponse) {
                         memberDescriptions.add(memberDescription);
                     }
                     final ConsumerGroupDescription consumerGroupDescription =
-                        new ConsumerGroupDescription(context.getGroupId(), protocolType.isEmpty(),
+                        new ConsumerGroupDescription(context.groupId(), protocolType.isEmpty(),
                             memberDescriptions,
                             describedGroup.protocolData(),
                             ConsumerGroupState.parse(describedGroup.groupState()),
-                            context.getNode().get(),
+                            context.node().get(),
                             authorizedOperations);
-                    context.getFuture().complete(consumerGroupDescription);
+                    context.future().complete(consumerGroupDescription);
+                }
+            }
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                context.future().completeExceptionally(throwable);
+            }
+        };
+    }
+
+    /**
+     * Returns a {@code Call} object to fetch the cluster metadata. Takes a List of Calls
+     * parameter to schedule actions that need to be taken using the metadata. The param is a Supplier
+     * so that it can be lazily created, so that it can use the results of the metadata call in its
+     * construction.
+     *
+     * @param <T> The type of return value of the KafkaFuture, like ListOffsetsResultInfo, etc.
+     * @param <O> The type of configuration option, like ListOffsetsOptions, etc
+     */
+    private <T, O extends AbstractOptions<O>> Call getMetadataCall(MetadataOperationContext<T, O> context,
+                                                                   Supplier<List<Call>> nextCalls) {
+        return new Call("metadata", context.deadline(), new LeastLoadedNodeProvider()) {
+            @Override
+            MetadataRequest.Builder createRequest(int timeoutMs) {
+                return new MetadataRequest.Builder(new MetadataRequestData()
+                    .setTopics(convertToMetadataRequestTopic(context.topics()))
+                    .setAllowAutoTopicCreation(false));
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                MetadataResponse response = (MetadataResponse) abstractResponse;
+                if (MetadataOperationContext.shouldRefreshMetadata(response)) {
+                    rescheduleMetadataTask(context, nextCalls);
```

Comment:
This is a bit subtle, but I think we want to raise the `InvalidMetadataException` rather than constructing a new `Call`. The problem is that we lose the retry bookkeeping which means these retries will not respect the backoff. By throwing the exception, we let the retry logic in `Call.fail` kick in. This would be consistent with the logic in  `getFindCoordinatorCall`.